### DISCUSSION TITLE
How to set wallpaper for each virtual desktop via D-Bus?

### DISCUSSION THREAD

**Discussion by iamhyc** [linuxdeepin/developer-center#3242] [Q&A | 问答板块] [answered]
Under the D-Bus interface "com.deepin.daemon.Appearance", there is only a read-only property for `WallpaperURIs`, but you could not set it.
So, is there a way to set the wallpaper for each virtual desktop?

**Comment by ghost** [comment]
> Under the D-Bus interface "com.deepin.daemon.Appearance", there is only a read-only property for `WallpaperURIs`, but you could not set it. So, is there a way to set the wallpaper for each virtual desktop?

Go on the virtual desktop where you want to set the wallpaper, right click from the mouse and select wallpaper from the menu. If you choose a wallpaper being on that desktop it will only set to it (not all desktops will have that wallpaper)

**Reply by iamhyc** [reply]
Yes, it can be set via mouse click. I just wonder whether there is an accessible API to do it.

I want to set the virtual desktop wallpaper via D-Bus or gsettings.

**Comment by iamhyc** [comment]
Hi @BLumia , could you invite some deepin developer to answer this?

**Reply by BLumia** [reply]
Sorry for the late reply. `SetCurrentWorkspaceBackground` under `com.deepin.wm` should be the one you're looking for :)

**Comment by BLumia** [comment, marked as answer]
Under `com.deepin.wm` there are a set of methods related to workspace background. You can use `SetCurrentWorkspaceBackground` for current workspace, and `SetWorkspaceBackground` for setting the background of the given virtual workspace index. You can use `d-feet` or `qdbusviewer` to find out other interfaces that you might interested :)

**Reply by iamhyc** [reply]
Got it. I did miss the `SetWorkspaceBackground` one.

**Comment by iamhyc** [comment]
@BLumia Just one more related question.
Why the first "workspace" is indexed with "1", instead of "0"?

It is not consistent with the "_NET_CURRENT_DESKTOP" value from x11. Is the idex 0 workspace reserved or for something else?

**Comment by iamhyc** [comment]
@BLumia Hi, I find `SetWorkspaceBackground` does not work correctly on 20.7.1.

I can check the wallpaper is already set via the "Get" method, but it does not take effect, i.e., the displayed wallpaper did not "refresh".

By the way, the wallpaper can be changed correctly via right click on the desktop.

Please help check if there is something wrong with the daemon. Thanks!

**Reply by BLumia** [reply]
Not sure, it should work imo. I saw you reported a bug (https://github.com/linuxdeepin/developer-center/issues/3505) and we'll check that out later :)
